modules/oe_migration_workflow/src/Plugin/migrate/process/SetWorkflowState.php

```diff
@@ -0,0 +1,259 @@
+use Drupal\migrate\Row;
+use Drupal\oe_migration\ValidConfigurableMigrationPluginInterface;
+use Drupal\workflows\WorkflowInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Sets the workflow state of an entity.
+ *
+ * If the input value is a valid workflow state, it will be returned
+ * untransformed. Else, if the workflow state is not set or is not a valid one,
+ * it will be set according to the status property, if the status is set to "1",
+ * then it will be set to the given "published_state", in any other case, it
+ * will be set to the given "unpublished_state".
+ *
+ * Compared to a static mapping, it can also handle cases where the source
+ * entity might not have a moderation state defined (NULL).
+ *
+ * Available configuration keys:
+ *   - workflow_config_name (mandatory): The destination workflow's
+ *     configuration name (string).
+ *     Default value: "workflows.workflow.oe_corporate_workflow".
+ *   - published_state (mandatory): The destination workflow's published state
+ *     (string).
+ *     Default value: "published".
+ *   - unpublished_state (mandatory): The destination workflow's unpublished
+ *     state (string).
+ *     Default value: "draft".
+ *
+ * @codingStandardsIgnoreStart
+ * Example with default configuration:
+ * @code
+ * process:
+ *   moderation_state:
+ *    - plugin: oe_migration_set_workflow_state
+ *      source: moderation_state
+ * @endcode
+ *
+ * Example with user defined configuration:
+ * @code
+ * process:
+ *   moderation_state:
+ *    - plugin: oe_migration_set_workflow_state
+ *      source: moderation_state
+ *      workflow_config_name: workflows.workflow.oe_corporate_workflow
+ *      published_state: published
+ *      unpublished_state: draft
+ * @endcode
+ * @codingStandardsIgnoreEnd
+ *
+ * @MigrateProcessPlugin(
+ *   id = "oe_migration_set_workflow_state"
+ * )
+ */
+class SetWorkflowState extends ProcessPluginBase implements ContainerFactoryPluginInterface, ValidConfigurableMigrationPluginInterface {
+
+  /**
+   * The configuration manager.
+   *
+   * @var \Drupal\Core\Config\ConfigManagerInterface
+   */
+  protected $configManager;
+
+  /**
+   * The entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * {@inheritdoc}
+   *
+   * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+   * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
+   * @throws \Drupal\migrate\MigrateException
+   */
+  public function __construct(
+    array $configuration,
+    $plugin_id,
+    $plugin_definition,
+    ConfigManagerInterface $config_manager,
+    EntityTypeManagerInterface $entity_type_manager) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    // The default configuration.
+    $this->configuration += [
+      'workflow_config_name' => 'workflows.workflow.oe_corporate_workflow',
```

Comment:
Done, really the workflow_config_name option already was set as mandatory, I don't know when or why it's configured with a _random_ default value.

Reply:
Thank you for this comment, it gives the opportunity to request clear guidelines from you (Core Team). Probably I misunderstood, I thought that inside an OE context we could use default values related to the environment. I specifically asked this question during our meeting with Francesco, Marco and Samuel in February 2020. Could you please give us clear directives? Thank you.

With that in mind,

- Since we're in an OpenEuropa and not in a Drupal.org context, the default values serve best the users inside this context and provide the default OE values, that can be overriden.
- You're absolutely right, it means nothing, you can as well use "test" as a default value. In any case, if one of the three values doesn't exist, an exception will be thrown and the migration will never start.
- We do not create any dependency, it's just a string that will be tested for its validity.
- In the same logic, maybe, we shouldn't use default values for draft or published. It's an assumption that another workflow will define these two states.

Thank you.

Reply:
OK, so to explain:

> Probably I misunderstood, I thought that inside an OE context we could use default values related to the environment.

That's not the point. You can use a default value from oe_ context, however, you need to put a dependency on that other component. Otherwise there is no connection. The two are completely decoupled. So if you wanna default to the corporate workflow, you need to composer.json depend on the editorial component. Otherwise, cannot make this assumption. That's the nature of decoupled components.

> Since we're in an OpenEuropa and not in a Drupal.org context, the default values serve best the users inside this context and provide the default OE values, that can be overriden.

Yes, but see my point above.

> In any case, if one of the three values doesn't exist, an exception will be thrown and the migration will never start.

yeah, of course, they need to be set in the migration, that's fine.

> In the same logic, maybe, we shouldn't use default values for draft or published. It's an assumption that another workflow will define these two states.

This is different. 1. It doesn't imply a value provided by another module. And 2, they are values that are gonna be like that in 90% of cases. Just like the "status" field name being "status". We cannot assume it is, but we can assume most are so we default to it. The 10% can change it. But it's different vs the corporate workflow assumption because that value couples it with another component.